astropy/nddata/nduncertainty.py

```diff
@@ -190,191 +276,483 @@ def parent_nddata(self):
-                    raise ValueError("array shape does not match "
-                                     "parent data shape")
-        self._array = value
+    def __getitem__(self, item):
+        """
+        Simple slicing is allowed and returns a reference *not* a copy. This
+        assimilates numpy slicing.
+        """
+        return self.__class__(self.array[item], unit=self.unit, copy=False)
 
-    def propagate_add(self, other_nddata, result_data):
+    def propagate(self, operation, other_nddata, result_data, correlation):
         """
-        Propagate uncertainties for addition.
+        Calculate the resulting uncertainty given an operation on the data.
 
         Parameters
         ----------
-        other_nddata : NDData instance
-            The data for the second other_nddata in a + b
-        result_data : `~numpy.ndarray` instance
-            The data array that is the result of the addition
+        operation: callable
+            The operation that is performed on the `NDData`. Supported are
+            `numpy.add`, `numpy.subtract`, `numpy.multiply` and
+            `numpy.true_divide`.
+
+        other_nddata: `NDData` instance
+            The second NDData in the arithmetic operation.
+
+        result_data: `~numpy.ndarray` or `~astropy.units.Quantity`
+            The result of the arithmetic operations. This saves some duplicate
+            calculations.
+
+        correlation: ``Number`` or `~numpy.ndarray`
+            The correlation (rho) is defined between the uncertainties in
+            sigma_AB = sigma_A * sigma_B * rho. A value of ``0`` means
+            uncorrelated operands.
 
         Returns
         -------
-        result_uncertainty : NDUncertainty instance
-            The resulting uncertainty
+        resulting_uncertainty: `NDUncertainty` instance
+            Another instance of the same `NDUncertainty` subclass containing
+            the uncertainty of the result.
 
         Raises
         ------
-        IncompatibleUncertaintiesException
-            Raised if the method does not know how to propagate the
-            uncertainties.
+        ValueError:
+            If the ``operation`` is not supported.
+
+        Notes
+        -----
+        This method at first tries to convert the ``uncertainty`` of the
+        other operand to a `NDUncertainty` subclass that is useable for
+        uncertainty propagation (this check and optional conversion is done
+        in :meth:`NDUncertainty._convert_uncertainty`) and then the
+        appropriate ``_propagate_*`` method for calculating the resulting
+        uncertainty is invoked. Afterwards this function wraps it into
+        another instance of `NDUncertainty` and returns it.
         """
+        # Check if the subclass supports correlation
+        if not self.supports_correlated:
+            if correlation != 0:
+                log.info("{0} does not support uncertainty propagation with "
```

Comment:
Just a note here: The line above makes real trouble. I have written the whole correlation support only for numbers and given an ``numpy.ndarray`` all the comparisons with 0 fail because ``ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()``

I think for now I'll change correlation to only support ``numbers`` and add ``np.ndarray`` correlations in a seperate PR. This could take some time, expecially writing tests.